content/post/high-availability-with-docker.md

```diff
@@ -0,0 +1,116 @@
+
+![Terminator](/images/high-availability-testing/terminal.png)
+
+That's all to do before getting our hands on Docker itself!
+
+
+Setup Docker Swarm on the cluster
+--------------------------------
+
+On latest HypriotOS, you'll have the latest Docker installed, so we can instantly start playing.
+
+On an arbitrary node of the cluster, run
+```
+docker swarm init
+```
+
+The output should look similar to that:
+```
+Swarm initialized: current node (node-master) is now a manager.
+
+To add a worker to this swarm, run the following command:
+
+    docker swarm join \
+    --token SWMTKN-1-1acm9qa7b0hmzz5v8t40d75v5fsgckeu2z5ds6ls0x7cny7l8p-307wqrc8756akpxjxls9abbbs \
+    [2a02:810d:8600:2a78:f6c6:d67d:6912:17ca]:2377
+
+To add a manager to this swarm, run 'docker swarm join-token manager' and follow the instructions.
+
+```
+
+As explained in this output, now you need to run the presented `docker swarm join ...` command on all other nodes, which you want to join the cluster.
+
+Finally, on the node at which you executed `docker swarm init`, check if all nodes of your cluster successfully formed a swarm:
+
+```
+docker node ls
+```
+
+If the terminal prints a list of all nodes of your cluster, the setup is all done! Forming a cluster cannot be easier, can it?
+
+
+Execution of tests
+---------------------
+My tests of Docker's ability to recover from failures comprise the following use cases:
+
+  - The ethernet interface of a random node (master or slave) got unavailable
+  - A random node completely rebooted
+  - A random node instantly crashed (unplug power source)
+
+I documented one of the test runs in the following screencast.
+
+<div align="center">
+<iframe src="https://player.vimeo.com/video/185361173" width="640" height="360" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe>
+<p><a href="https://vimeo.com/185361173">Testing High Availability Docker Swarm Mode</a> from <a href="https://vimeo.com/user54109827">Mathias Renner</a> on <a href="https://vimeo.com">Vimeo</a>.</p>
+</div>
+
+Results
+------------
+As shown in the screencast, Docker is able to recover from failure of the ethernet interface. After testing the other use cases later on, I can confirm that Docker recovers flawlessly from a reboot and crash as well. This holds for the slave and the master node, which is remarkable!
+
+
+More to come soon
+------------------
+This post is only a small chunk of the data I gathered during the tests. The next option to get the details is during my talk at the [HighLoad++ Conference](http://highload.co/) in Moskow, Russia. I'd be happy if you can make it there!
+
+Also, this is not the end of the story,kontakt@pringuin.de of course. Thorough testing requires also measuring incoming requests from an external load tester to the cluster while a failure occurs. After some research, I have not found any evidence that someone has ever performed that tests (please correct me if I'm wrong!).
```

Comment:
This sentence is all wrong: Also, this is not the end of the story,kontakt@pringuin.de of course.